wrappers/cpp/acrawriter.hpp

```diff
@@ -0,0 +1,124 @@
+#include <themis/themis.h>
+
+// themispp should be installed as system library (run `themispp_install` from Themis code folder)
+#include <themispp/secure_cell.hpp>
+#include <themispp/secure_message.hpp>
+#include <themispp/secure_keygen.hpp>
+
+namespace acrawriter {
+
+    using namespace themispp;
+    using namespace std;
+
+    class acrawriter {
+
+    public:
+        typedef vector<uint8_t> data;
+        typedef vector<uint8_t> acrastruct;
+
+        acrastruct create_acrastruct(const data &message, const data &zone_public_key, const data &zone_id) {
+          return create_and_pack_acrastruct(message, zone_public_key, zone_id);
+        }
+
+        acrastruct create_acrastruct(const data &message, const data &public_key) {
+          return create_and_pack_acrastruct(message, public_key);
+        }
+
+    private:
+        // because it's always 32
+        const uint8_t symmetric_key_length = 32;
+
+        // AcraStruct header format has eight ' symbols, 34 is ascii code of '
+        const uint8_t acrastruct_header_byte = 34;
+
+        // private
+        acrastruct create_and_pack_acrastruct(const data &message, const data &public_key, const data &zone_id = data()) {
+          // 1. generate EC keypair
+          secure_key_pair_generator_t<EC> key_pair_generator;
+          data temp_private_key = key_pair_generator.get_priv();
+          data temp_public_key = key_pair_generator.get_pub();
+
+          // 2. generate random symm key with `symmetric_key_length` size
+          using random_bytes_engine = independent_bits_engine<default_random_engine, CHAR_BIT, uint8_t>;
+          random_bytes_engine rbe;
+          vector<uint8_t> symmertic_key(symmetric_key_length);
+          generate(begin(symmertic_key), end(symmertic_key), ref(rbe));
+
+          // 3. encrypt random symmetric key using asymmetric encryption with random private key and acra/zone public key
+          secure_message_t secure_message(temp_private_key, public_key);
+          data encrypted_symm_key = secure_message.encrypt(symmertic_key);
+
+          // 4. encrypt payload using symmetric encryption and random symm key
+          secure_cell_seal_t secure_cell(symmertic_key);
+          data encrypted_message;
+
+          if (zone_id.empty()) {
+            encrypted_message = secure_cell.encrypt(message);
+          } else {
+            encrypted_message = secure_cell.encrypt(message, zone_id);
+          }
+
+          // pack into array because I don't know how to push uint64_t into vector<uint8_t>
+          uint64_t em_length = encrypted_message.size();
+          uint8_t em_length_array[8];
+          memcpy(em_length_array, &em_length, sizeof(em_length));
+
+          // convert encrypted data length to little endian
+          if (is_big_endian()) {
+            reverse(begin(em_length_array), end(em_length_array));
+          }
+
+          // zeroing symmertic key
+          fill(symmertic_key.begin(), symmertic_key.end(), 0);
```

Comment:
maybe would be good to fill with zeroes `temp_private_key` too?

Reply:
agree